`python`: add support for typed custom options

alltilla opened the pull request:
This applies for `python` source, `python-fetcher` source, `python` destination, `python` parser and `python-http-header` inner destination.

Example config:
```
python(
  class("TestClass")
  options(
    "string_option" => "example_string"
    "bool_option" => yes  # becomes True in the python code
    "integer_option" => 123456789
    "double_option" => 123.456789
    "string_list_option" => ["string1", "string2", "string3"]
  )
);
```

Signed-off-by: Attila Szakacs <szakacs.attila96@gmail.com>

bazsi commented:
This sounds like a great idea. Thanks for creating this

bazsi commented on modules/python/python-types.c:
```diff
@@ -34,15 +34,23 @@
 py_bytes_from_string(const char *value, gssize len)
 {
+  PyGILState_STATE gstate = PyGILState_Ensure();
+
```
I am a bit concerned if this has a runtime overhead for the normal case, e.g. during production, when the gil is already taken. If this is non trivial then I think we should create wrappers instead.

alltilla replied:
This got solved while introducing the `PythonOption` class.

bazsi commented on modules/python/python-grammar.ym:
```diff
@@ -135,13 +135,7 @@ python_dd_custom_options
-          free($2);
-        }
-	| string LL_ARROW string_or_number
+        : string optional_arrow string_or_number
```
This also removes the actual setting of options, I guess it is readded by a subsequent commit. Would be nicer with not removing that.

alltilla replied:
Actually it does not, just one of the two setter rule become unused. However, this was solved differently with the `PythonOptions` class, so it is kinda "fixed".

bazsi commented on modules/python/python-dest.c:
```diff
@@ -86,11 +86,12 @@ python_dd_set_class(LogDriver *d, gchar *class_name)
 
 void
-python_dd_set_option(LogDriver *d, gchar *key, gchar *value)
+python_dd_add_option(LogDriver *d, const gchar *name, PyObject *value)
```
Can't we encapsulate the hashtable and related functions into a PythonOptions (or similarly named) structure which the python components would just contain?

alltilla replied:
My first implementation was like you have mentioned, but half way through it, I felt like it had too much boilerplate. However, I finished that implementation now and it is okay, I guess. Now we have 2 implementations, so we can decide :)

bazsi commented on tests/light/functional_tests/source_drivers/python_fetcher/test_python_fetcher_custom_options.py:
```diff
@@ -0,0 +1,60 @@
+#############################################################################
+
+
+def test_python_fetcher_custom_options(config, syslog_ng):
```
We wouldn't need to test all python components for all types, if we just merged the implementation into one. Then a simple happy path that the option stuff is working would be enough in my opinion.

Test execution time is a scarce resource too.

alltilla replied:
I would really like to test all of these, as even with the new `PythonOptions` class there are still some code duplication. Also the python confgen has a completely different implementation.

I added python confgen tests, and merged all of these together to one light TC, which runs for 0.22 seconds, so I guess it is good enough.

alltilla commented:
@bazsi I have rebased this to master and fixed your comments. I feel like this implementation is good enough on the boilerplate front, but we can still carve from it if you find it necessary. Could you kindly take a look at it? Thanks!

alltilla commented:
Also added template support.

alltilla commented:
> Also added template support. Not sure how to differentiate it from the simple string type in the config. I chose `LogTemplate()`, which is the name of the python class, but we could use `template()` or something similar.
> 
> I like the idea, that we set the right handside with a python-like syntax, like with [] for the string list and here with LogTemplate. The only exception is yes and no, but we could make it, so only here, True and False is allowed additionally to yes and no, so we can write confgens and scls easily with yes and no, and we can also provide python like syntax with True and False.

Implemented that in a separate commit. We can drop it, if we want to do this in a different way.

bazsi commented on modules/python/python-types.c:
```diff
@@ -124,6 +124,29 @@ py_list_from_list(const gchar *list, gssize list_len)
+      PyObject *py_string = py_string_from_string(string, -1);
+      if (!py_string || PyList_Append(obj, py_string) != 0)
+        {
+          Py_DECREF(obj);
```
if PyList_Append() fails we leak py_string

alltilla replied:
Thanks, fixed!

bazsi commented on modules/python/python-types.c:
```diff
@@ -345,6 +368,31 @@ py_list_to_list(PyObject *obj, GString *list)
+
+      const gchar *element_as_str;
+      if (!py_bytes_or_string_to_string(element, &element_as_str))
+        return FALSE;
```
we should probably clean up the partially built string_list in this case.

alltilla replied:
Thanks, fixed!

bazsi commented on modules/python/python-grammar.ym:
```diff
@@ -330,6 +334,55 @@ python_parser_loaders
+          : string optional_arrow string
+            {
+              PythonOption *option = python_option_string_new($1, $3);
+              python_options_add_option(last_python_options, option);
```
this rule could return the PythonOption instance in $$ so the reference of this rule could perform python_options_add_option(), avoiding some of the duplication here.

bazsi replied:
Also, python_options_add_option() will clone its argument, so "option" itself is leaked here.

I think we could get away with consuming the reference passed as argument, so no need for the clone. Maybe instead of clone we would only need refcounting?

alltilla replied:
I think I will just get rid of the cloning as a whole. It is only used in the grammar, and it was only done to keep the ownerships "cleaner", but I don't think it served its purpose well.

alltilla replied:
Done.

bazsi reviewed (commented):
I had some smaller issues posted as review comments, otherwise this looks good. The LogTemplate functionality is not present the PR but is mentioned in the summary.

alltilla commented:
> The LogTemplate functionality is not present the PR but is mentioned in the summary.

D'oh, that's what I get for developing this on 2 computers... I will dig up the LogTemplate functionality and force push this PR.

bazsi commented on modules/python/python-options.c:
```diff
@@ -268,6 +269,54 @@ python_option_string_list_new(const gchar *name, const GList *value)
+  PyObject *template_str = py_string_from_string(self->value, -1);
+  PyObject *args = PyTuple_Pack(1, template_str);
+  PyObject *py_template = PyObject_Call((PyObject *) &py_log_template_type, args, NULL);
+  Py_DECREF(template_str);
```
we might need to handle exceptions as all of these functions can return NULL in case of failure.

alltilla replied:
Nice catch! I have handled exceptions by logging an error, but still continuing to build the rest of the `dict()`.

alltilla commented:
Rebased to master, resolved the merge conflict.

MrAnno commented on modules/python/python-options.c:
```diff
@@ -233,6 +234,51 @@ python_option_string_list_new(const gchar *name, const GList *value)
+    return NULL;
+
+  PyObject *args = PyTuple_Pack(1, template_str);
+  PyObject *py_template = PyObject_Call((PyObject *) &py_log_template_type, args, NULL);
```
Why don't we just call `py_log_template_new()` here?

MrAnno replied:
Oops. Because it is not implemented, only the declaration exists.

MrAnno commented on news/feature-4354.md:
```diff
@@ -0,0 +1,19 @@
+  options(
+    "string_option" => "example_string"
+    "bool_option" => True  # supported values are: True, False, yes, no
+    "integer_option" => 123456789
```
This PR introduces a little breaking change.
In the case of numbers and previously unquoted strings (true/false, yes/no), the Python-side representation changes.

If it would be difficult to keep the original behavior for old config versions, then I think we should mention it in the NEWS file.

alltilla replied:
Nice catch! I have added a runtime warning and a **Breaking change** sentence to the news file entry.

MrAnno replied:
Thanks.

MrAnno commented on modules/python/python-options.c:
```diff
@@ -0,0 +1,336 @@
+        if (!value)
+          continue;
+
+        PyDict_SetItemString(py_dict, name, value);
```
This and the constructor call may fail. I don't know how much this actually happens. Probably in OOM cases.

alltilla replied:
Thanks, I have handled those errors. In case of ctor fail, we can only return NULL, in case of set failure I have added an error message and continued.

MrAnno reviewed (commented):
LGTM.
